include/zephyr/bluetooth/l2cap.h

```diff
@@ -147,7 +147,12 @@ struct bt_l2cap_le_endpoint {
 	uint16_t				mtu;
 	/** Endpoint Maximum PDU payload Size */
 	uint16_t				mps;
-	/** Endpoint initial credits */
+	/** Endpoint initial credits
+	 *
+	 * The user should allocate enough credits to be able to transfer a full
+	 * SDU. 1 credit = 1 PDU of MPS length. Example:
+	 * SDU = 2000 bytes, MPS = 77 -> init_credits >= 26.
```

Comment:
This is not enough to ensure successful transport. The remote may choose to underutilize MPS. The comment we are removing below hinted at exactly this.

https://github.com/zephyrproject-rtos/zephyr/blob/bb6fa5707f77d72307008405f19dd83247405392/subsys/bluetooth/host/l2cap.c#L2460-L2463

We need a contract with the application:

When the channel connected event handlers complete, `bt_l2cap_chan_ops.alloc_seg` shall be ready give out `init_credits` net_bufs. When the net_buf is destructor returns, a net_buf must be available from `alloc_seg` again. 

`bt_l2cap_chan_ops.alloc_buf` shall be ready give out one buffer after..
 - channel connected event handlers complete
 - when the channel's `recv` returns other than `-EINPROGRESS`
 - when the channel's `bt_l2cap_chan_recv_complete` is invoked

Contract is valid until the 'disconnected' event.

Additionally: (deprecated) If `bt_l2cap_chan_ops.alloc_seg` is NULL, then `alloc_buf` must in addition give out the buffers demanded of `bt_l2cap_chan_ops.alloc_seg`.

We make sure the data in a `seg` is copied to the `buf`, and `seg` is destroyed before we accept more ACL data from that conn.

Reply:
I realize the `-EINPROGRESS` situation is not completely fleshed out. I suppose we can demand `init_credits` buffers of `alloc_buf` if `-EINPROGRESS` is used. Seems wasteful though.

Reply:
When `-EINPROGRESS`, the `seg` start running out at the same rate as credits. This is exactly what we want. When `bt_l2cap_chan_recv_complete` is invoked, we can immediately grab a new `alloc_buf`, copy and destroy all received `segs`, and send back more credits.

Reply:
@alwa-nordic I think you misunderstood how `.recv` works. It is only ever called after receiving a complete SDU.

This is the whole point of the PR, where the user basically has no levers it can pull to send more credits _when receiving PDUs_. It can only withhold credits by returning `-EINPROGRESS` after having received a full SDU.

Because this is the API that we have, IMO the credits should be an implementation detail, and should be managed by the host and not by the users at all. 

Even initial_credits is not a good idea with this API; I think the application should rather inform the stack of its SDU buffer size instead.